src/core/tiledmesh/qgscesiumtilesdataprovider.cpp

```diff
@@ -28,26 +28,256 @@
+  }
+
+  node->setGeometricError( json["geometricError"].get< double >() );
+  if ( json["refine"] == "ADD" )
+    node->setRefinementProcess( Qgis::TileRefinementProcess::Additive );
+  else if ( json["refine"] == "REPLACE" )
+    node->setRefinementProcess( Qgis::TileRefinementProcess::Replacement );
+
+
+  if ( json.contains( "content" ) && !json["content"].is_null() )
+  {
+    const auto &contentJson = json["content"];
+
+    // sometimes URI, sometimes URL...
+    if ( contentJson.contains( "uri" ) && !contentJson["uri"].is_null() )
+    {
+      node->setContentUri( mRootPath + '/' + QString::fromStdString( contentJson["uri"].get<std::string>() ) );
+    }
+    else if ( contentJson.contains( "url" ) && !contentJson["url"].is_null() )
+    {
+      node->setContentUri( mRootPath + '/' + QString::fromStdString( contentJson["url"].get<std::string>() ) );
+    }
+  }
+
+  if ( json.contains( "transform" ) && !json["transform"].is_null() )
+  {
+    const auto &transformJson = json["transform"];
+    QgsMatrix4x4 transform;
+    double *ptr = transform.data();
+    for ( int i = 0; i < 16; ++i )
+      ptr[i] = transformJson[i].get<double>();
+
+    if ( parentTransform )
+    {
+      node->setTransform( *parentTransform * transform );
+    }
+    else
+    {
+      node->setTransform( transform );
+    }
+  }
+  else if ( parentTransform )
+  {
+    node->setTransform( *parentTransform );
+  }
+
+  return node;
+}
+
+const QgsTiledMeshNode *QgsCesiumTiledMeshIndex::rootNode() const
+{
+  return mRootNode.get();
+}
+
+QgsTiledMeshNode *QgsCesiumTiledMeshIndex::getNodes( const QgsTiledMeshRequest &request )
+{
+  if ( !mRootNode )
+    return nullptr;
+
+  std::function< QgsTiledMeshNode * ( const json &nodeJson, QgsTiledMeshNode * )> traverseNode;
+  traverseNode = [&request, &traverseNode, this]( const json & nodeJson, QgsTiledMeshNode * parent ) -> QgsTiledMeshNode *
+  {
+    std::unique_ptr< QgsTiledMeshNode > newNode;
+    if ( parent )
+    {
+      const QgsMatrix4x4 parentTransform = parent->transform();
+      newNode = nodeFromJson( nodeJson, &parentTransform );
+    }
+    else
+    {
+      newNode = nodeFromJson( nodeJson, nullptr );
+    }
+
+    if ( parent )
+      parent->addChild( newNode.get() );
+
+    // TODO -- if node has content json, fetch it now and parse
+
+    if ( request.requiredGeometricError() <= 0 || newNode->geometricError() <= 0 || newNode->geometricError() > request.requiredGeometricError() )
+    {
+      // haven't traversed deep enough down this node, we need to explore children
+      if ( nodeJson.contains( "children" ) )
+      {
+        for ( const auto &childJson : nodeJson["children"] )
+        {
+          if ( request.feedback() && request.feedback()->isCanceled() )
+            break;
+
+          traverseNode( childJson, newNode.get() );
+        }
+      }
+    }
+
+    return newNode.release();
+  };
+
+  return traverseNode( mTileset[ "root" ], nullptr );
+}
```

Comment:
I am a bit confused why we would want to do this approach - to keep the hierarchy as `json` object and ad-hoc parse data from it as needed? It seems to me it would be much simpler to just parse the JSON once, make our in-memory hierarchy of QgsTiledMeshNode objects - but I may be missing something :slightly_smiling_face:  Also when we start to deal with a hierarchy of JSON files, the traverse logic may become more complex...